php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +63,334 @@ class StudentCompetency extends \ActiveRecord
         ]
     ];
 
-    public static function isCurrentLevelComplete($Student, $Competency)
+    public static $dynamicFields = [
+        'demonstrationsLogged' => [
+            'getter' => 'getDemonstrationsLogged'
+        ],
+        'demonstrationsComplete' => [
+            'getter' => 'getDemonstrationsComplete'
+        ],
+        'demonstrationsAverage' => [
+            'getter' => 'getDemonstrationsAverage'
+        ],
+        'demonstrationData' => [
+            'getter' => 'getDemonstrationData'
+        ],
+        'effectiveDemonstrationsData' => [
+            'getter' => 'getEffectiveDemonstrationsData'
+        ]
+
+    ];
+
+    public function calculateStartingRating($autoSave = true)
     {
-        $completion = $Competency->getCompletionForStudent($Student);
+        $validSkills = DB::allRecords(
```

Comment:
@nbey read but ignore most of my other feedback in this function, instead start it off with something like this:

```php
$ratings = DB::valuesTable(
    'SkillID',
    'Rating',
    '
    SELECT S.ID AS SkillID,
        IFNULL(
            JSON_EXTRACT(DemonstrationsRequired, \'$."%4$u"\'),
            JSON_EXTRACT(DemonstrationsRequired, "$.default")
        ) AS DemonstrationsRequirements,
        (
            SELECT DS.DemonstratedLevel
              FROM `%3$s` DS
              JOIN `%2$s` D
                ON D.ID = DS.DemonstrationID
             WHERE DS.SkillID = S.ID
               AND DS.TargetLevel = %4$u
               AND DS.DemonstratedLevel > 0
               AND D.StudentID = %5$u
             ORDER BY D.Demonstrated, D.ID
             LIMIT 1
        ) AS Rating
    FROM `%1$s` S
    WHERE S.CompetencyID = %6$u
    HAVING DemonstrationsRequirements > 0
    ',
    [
        Skill::$tableName, // %1$s
        Demonstration::$tableName, // %2$s
        DemonstrationSkill::$tableName, // %3$s
        'Level' => 9, // %4$u
        'StudentID' => 23, // %5$u
        'CompetencyID' => 1 // %6$u
    ]
)
```

This requires CBL-319 be done to eliminate the string->int conversion